SCClassLibrary/Common/Control/Server.sc

```diff
@@ -348,23 +370,43 @@ Server {
 	}
 
 	initTree {
-		nodeAllocator = NodeIDAllocator(clientID, options.initialNodeID);
-		this.sendMsg("/g_new", 1, 0, 0);
+		this.newNodeAllocators;
+		this.sendMsg("/g_new", this.defaultGroup.nodeID, 0, 0);
 		tree.value(this);
 		ServerTree.run(this);
 	}
 
 	/* id allocators */
 
+	// private, called from server notify response with next free clientID
 	clientID_ { |val|
+		var failstr = "Server % couldn't set client id to: % - %. clientID is still %.";
+
+		if(val == clientID) {
+			// no need to change
+			^this
+		};
 		if(val.isInteger.not) {
-			"Server % couldn't set client id to: %".format(name, val.asCompileString).warn;
+			failstr.format(name, val.cs, "not an Integer", clientID).warn;
 			^this
 		};
-		if(clientID != val) {
-			clientID = val;
-			this.newAllocators;
-		}
+		if (val < 0) {
+			failstr.format(name, val.cs, "less than minimum 0", clientID).warn;
+			^this
+		};
+		if (val >= this.numClients) {
+			failstr.format(name,
+				val.cs,
+				"greater than server.numClients % minus 1 allows".format(this.numClients),
+				clientID
+			).warn;
+			^this
+		};
+
+		"% : setting clientID to %.\n".postf(this, val);
+		clientID = val;
+		this.newAllocators;
+		this.initTree;
```

Comment:
`initTree` will also reastart all synths etc. that have been registered for startup in `ServerTree` - won't it? So won't we get duplicate actions here unless we cleanup before this?

Reply:
Setting the clientID while there are still synths running works, but may break and you might get duplicate node ids in a multi-client setup. What to do about this? At least we need to document this.

```
a = Synth(\default);
s.clientID = 2;
```
the synth is still there, but now has "wrong" node id.

Reply:
> initTree will also restart all synths etc. that have been registered for startup in ServerTree - won't it? So won't we get duplicate actions here unless we cleanup before this?

yes, this needs to be cleaner: 
initTree should only happen after the server notify response, not already in bootInit.
Then we always have a valid default, and do not make one that may be changed too early.

---
> Setting the clientID while there are still synths running works, but may break and you might get duplicate node ids in a multi-client setup. What to do about this? At least we need to document this.

Yes, would be good to lock it, as @muellmusik also proposed.
ATM StatusWatcher needs clientID_ to set it one way or another...
actually, one could move the notify response handlers to Server, then it would be inaccessible
This would be Server -> prHandleNotifyFailString and prHandleClientLoginInfoFromServer.
yes?

> 
> ```
> a = Synth(\default);
> s.clientID = 2;
> ```
> the synth is still there, but now has "wrong" node id.

yes, and this is actually fine - if you've kept the synth accessible, you can still use it after a change of clientID; when you create new ones, they will be allocated in the new range.
But when we do the above, clientID will be inaccessible except from the OSC response.

thanks!

Reply:
actually the safety checks on clientID are still useful, and make sense in a separate method. So, we can tell users not to change clientID by hand unless they really know what they are doing.

Reply:
>ATM StatusWatcher needs clientID_ to set it one way or another...
actually, one could move the notify response handlers to Server, then it would be inaccessible
This would be Server -> prHandleNotifyFailString and prHandleClientLoginInfoFromServer.
yes?

I see, you mean that `ServerStatusWatcher` is not necessarily the right place to organize the notification request handling. Makes sense.

That means, by moving it back to `Server`, `sendNotifyRequest` could handle the `clientID` setting from within the server, so that it couldn't be set from the outside?

I'd even go so far as to propose a class `ServerNotifier` which handles it, and the privacy of the setter message `clientID_()` could be kept by a redirect to a `setClientID`. But maybe I miss what you are proposing.

For the current PR, this refactoring isn't necessary yet.

Reply:
I meant, ServerStatusWatcher can still set up the oscfuncs, which receive the response, 
and Server:prHandleNotifyResponse and prHandleNotifyFailResponse get the raw message from the OSCFunc, and do the rest. Then we could remove clientID_ for encapsulation, and only set clientID either when creating a new server, or via osc message from scsynth.